core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -0,0 +1,211 @@
+                                       initialPrincipalName: String)
+
+class BrokerToControllerRequestThread(networkClient: KafkaClient,
+                                      metadataUpdater: ManualMetadataUpdater,
+                                      requestQueue: BlockingQueue[BrokerToControllerQueueItem],
+                                      metadataCache: kafka.server.MetadataCache,
+                                      config: KafkaConfig,
+                                      listenerName: ListenerName,
+                                      time: Time,
+                                      threadName: String)
+  extends InterBrokerSendThread(threadName, networkClient, time, isInterruptible = false) {
+
+  private var activeController: Option[Node] = None
+
+  override def requestTimeoutMs: Int = config.controllerSocketTimeoutMs
+
+  override def generateRequests(): Iterable[RequestAndCompletionHandler] = {
+    val requestsToSend = new mutable.Queue[RequestAndCompletionHandler]
+    if (requestQueue.peek() != null) {
+      val topRequest = requestQueue.take()
+
+      val request = RequestAndCompletionHandler(
+        activeController.get,
+        topRequest.request,
+        handleResponse(topRequest),
+        topRequest.initialPrincipalName)
+      requestsToSend.enqueue(request)
+    }
+    requestsToSend
+  }
+
+  private[server] def handleResponse(request: BrokerToControllerQueueItem)(response: ClientResponse): Unit = {
+    if (response.wasDisconnected()) {
+      activeController = None
+      requestQueue.put(request)
+    } else if (response.responseBody().errorCounts().containsKey(Errors.NOT_CONTROLLER)) {
+      // just close the controller connection and wait for metadata cache update in doWork
+      networkClient.close(activeController.get.idString)
+      activeController = None
+      requestQueue.put(request)
+    } else {
+      request.callback.onComplete(response)
+    }
+  }
+
+  private[server] def backoff(): Unit = pause(100, TimeUnit.MILLISECONDS)
+
+  override def doWork(): Unit = {
+    try {
+      if (activeController.isDefined) {
+        super.doWork()
+      } else {
+        debug("Controller isn't cached, looking for local metadata changes")
+        val controllerOpt = metadataCache.getControllerId.flatMap(metadataCache.getAliveBroker)
+        if (controllerOpt.isDefined) {
+          if (activeController.isEmpty || activeController.exists(_.id != controllerOpt.get.id))
+            info(s"Recorded new controller, from now on will use broker ${controllerOpt.get.id}")
+          activeController = Option(controllerOpt.get.node(listenerName))
+          metadataUpdater.setNodes(metadataCache.getAliveBrokers.map(_.node(listenerName)).asJava)
+        } else {
+          // need to backoff to avoid tight loops
+          debug("No controller defined in metadata cache, retrying after backoff")
+          backoff()
+        }
+      }
+    } catch {
+      case e: Exception =>
```

Comment:
What exceptions can actually be captured here? It looks like InterBrokerSendThread will kill the JVM if it encounters any Throwable by throwing a FatalExitError (which won't be caught as an Exception). 

It looks like IBST will handle disconnects by setting the `disconnected` flag in the ClientResponse (which we're checking up at L165)

Reply:
I'm neutral to do a full exception catch here, if the underlying thread was trying to kill the entire broker proactively. The path for undefined controller in the try block should not trigger any exception in theory, so it makes sense to just throw IMHO.

Reply:
Removed the try-catch